### PR TITLE
T87 - Improve group scope dropdown ui

### PR DESCRIPTION
**Pull request by robertknight**
This is the initial cut of the improved group link sharing UI.

Remaining tasks:
 * Implement suitable tests
 * Clarify design when there are many groups in the list. Currently it will grow ad-infinitum

group-share-ui-update

### CODE REVIEW

**Comment by dwhly** [comment]
Nice!

I kind of like the copy URL inlined in the card.  Agree that dealing w/ long lists is important.

**Comment by robertknight** [comment]
Perhaps we should do the same for creating new groups as well. Currently the process of creating a new group looks like this:

1. Click 'New group'
2. That opens up a very spartan tab that asks for a single piece of data (the group name) which shows a 'Create Group' button.
3. Clicking 'Create Group' creates the group and gives you a link, but when you go back to the original tab to make use of the new group, it doesn't appear until you refresh the page.

Inlining the 'Create Group' action into the above menu would make the process slicker. The potential downside is that you wouldn't have room to offer any explanations about groups in that menu, but would have to link out to a separate page for that.

**Comment by dwhly** [comment]
> The potential downside is that you wouldn't have room to offer any explanations about groups in that menu, but would have to link out to a separate page for that.

That in and of itself might be a DQ.

**Comment by dwhly** [comment]
So, been going back over @conordelahunty's designs after my somewhat off the cuff remarks earlier.  Follow up thoughts:

1- We should stick to a design process here, and respect our designer (not that you were intending otherwise!) -- giving him feedback and letting him evolve the design in partnership with us, but ultimately having him own it.

2- On this particular suggestion, from a consistency POV, it's definitely important to maintain consistency between creating and sharing a group.  Also, the designs as made provide for the social sharing icons in addition to just the form field, and as you suggest adding the rational / description for a group is difficult in the smaller space and something we've discussed before.  (Card for it: https://trello.com/c/DByfGdQt/264-add-descriptions-for-groups)

**Comment by robertknight** [comment]
> So, been going back over @conordelahunty's designs after my somewhat off the cuff remarks earlier.

To clarify, there have been several rounds of design, the one I'm working to is https://trello.com/c/08dCx7gd/87-group-scope-dropdown-ui-updates which came out of the meeting on Tuesday this week.

> 1- We should stick to a design process here, and respect our designer (not that you were intending otherwise!) -

Yes, of course. What is actually implemented in this PR follows the [original spec](https://trello.com/c/08dCx7gd/87-group-scope-dropdown-ui-updates) (at least, assuming I've correctly understood all aspects of it).

**Comment by dwhly** [comment]
Ack. Forgive, I must have been looking at the completely wrong card. Apologies!  Feeling like an idiot now.

File: h/static/scripts/directive/group-list.js [hunk 1]
```diff
@@ -1,17 +1,64 @@
...
  * @restrict AE
  * @description Displays a list of groups of which the user is a member.
  */
-
 // @ngInject
 module.exports = function (groups) {
   return {
-    link: function (scope, elem, attrs) {
-      scope.groups = groups;
+    controller: ['$scope', GroupsListController],
```
Comment (nickstenning) [inline h/static/scripts/directive/group-list.js]: If you add an `// @ngInject` comment before the definition of `GroupsListController`, the array-form function definition isn't necessary.

File: h/static/scripts/directive/group-list.js [hunk 2]
```diff
@@ -1,17 +1,64 @@
...
+    link: function ($scope, elem, attrs) {
+      $scope.groups = groups;
+
+      // set the base URI used later to construct the sharing
+      // link for the group
+      $scope.baseURI = elem[0].ownerDocument.baseURI;
```
Comment (nickstenning) [inline h/static/scripts/directive/group-list.js]: You shouldn't need to construct group URLs. The data sent from the server for groups [includes a prerendered URL](https://github.com/hypothesis/h/blob/4a324db9c3495e71a4a1c3752617d7b9e957e195/h/session.py#L40-L42).

File: h/static/styles/app.scss [hunk 3]
```diff
@@ -82,27 +82,82 @@ ol {
...
-}
+    .group-name {
+      overflow: hidden;
+      text-overflow: ellipsis;
+      width: $group-list-width - 30px;
+    }
+  }
+
+  .group-item {
+    display: flex;
```
Comment (nickstenning) [inline h/static/styles/app.scss]: Nice :)

File: h/templates/app.html.jinja2 [hunk 4]
```diff
@@ -29,10 +29,8 @@
            on-clear="search.clear()"></div>
       <!-- / Searchbar -->
 
-      <div class="group-list"
-           ng-if="auth.user && feature('groups')",
-           group-list="">
-      </div>
+      <group-list class="group-list" ng-if="auth.user && feature('groups')">
```
Comment (nickstenning) [inline h/templates/app.html.jinja2]: :+1: To element directives.

**Comment by nickstenning** [comment]
Okay, this is looking really good. Just a couple of notes in the code, and a couple of notes on visuals:

- Comparing with @conordelahunty's mockup, I think the line-spacing between the group name and the "Invite others" link needs tightening up a bit.
- The "New group" link should have a lowercase 'g', and the "add" icon needs a heavier weight to match up with the other icons in the dropdown.

**Comment by conordelahunty** [comment]
@robertknight (catching up on your comments from a couple of days ago, sorry) I wanted to keep the group creation out of the dropdown for exactly the reason you mention, not enough space to explain what to do. I'd like it to be more inline with the group invitation page, I will have to provide you with an updated design for that.

@dwhly Sorry about the minor confusion, we made a couple of quick calls on things last week when we were working through the details and I didn't flag them with you.

**Comment by conordelahunty** [comment]
@robertknight Looking good btw. How would you like me to feedback on visual bugs? 
We can do it on a case by case or do a sweep of them later?

**Comment by robertknight** [comment]
> How would you like me to feedback on visual bugs? 

Annotate a screenshot of the bug with a big arrow pointing out the error :)

>  We can do it on a case by case or do a sweep of them later?

I prefer feedback earlier rather than later, so I think case-by-case is preferred.

**Comment by robertknight** [comment]
scope-dropdown-impl

**Comment by dwhly** [comment]
> Sorry about the minor confusion, we made a couple of quick calls on things last week when we were working through the details and I didn't flag them with you.

No worries.

So, we don't need to necessarily address now, but I *do* think we should generally maintain consistency across our similar interfaces.   As @robertknight observes, the invite and create interfaces are two sides of the same coin.

For instance, the group create dialog gives you a URL to copy-paste and then a set of social shortcuts which create stub emails/tweets/fb posts for you.  On mobile this can be a particular advantage (by decreasing the number of compound interactions required to do copy, and then hunt for a given app to paste the clipboard to).  If that's a useful paradigm, we should carry it throughout?

(Personally I'm not bothered by the dedicated card for group creation.  It's the same number of clicks to reveal / dismiss it as the invite UI?)

**Comment by conordelahunty** [comment]
@robertknight @dwhly Agreed on the consistency. Take a look at a revised design of group creation below. This last step could also be how we share the group.

groups-creation-process

**Comment by nickstenning** [comment]
LGTM. @conordelahunty?

File: h/static/scripts/directive/group-list.js [hunk 5]
```diff
@@ -1,17 +1,61 @@
...
 module.exports = function (groups) {
   return {
-    link: function (scope, elem, attrs) {
-      scope.groups = groups;
+    controller: GroupsListController,
+    link: function ($scope, elem, attrs) {
+      $scope.groups = groups;
+
+      $scope.createNewGroup = function() {
+        window.open('/groups/new', '_blank');
```
Comment (nickstenning) [inline h/static/scripts/directive/group-list.js]: I'm not sure I understand why this became JS rather than an href?

**Comment by conordelahunty** [comment]
@robertknight As for the shot you pasted above - move the invite link closer to the title by 3-4px. Also, can you reduce the line height of the "You can invite other people to..." copy. And add a little more top and bottom padding to the "+ New group" cell.

dropdown-feedback

**Comment by dwhly** [comment]
We should be consistent in the share options we support.   From the share dialog for pages, we have:

![screen shot 2015-09-21 at 9 00 13 am]

(https://cloud.githubusercontent.com/assets/1413074/9997276/1f03b474-6040-11e5-99b7-d07cabd0394c.png)

Also, should we be consistent in the UI style between these various actions?  I'm sensing that there's a potential shift to a new aesthetic paradigm here, but it's not complete.  Should we keep the current style until we're ready to make a uniform shift?

File: h/static/scripts/directive/group-list.js (hunk 5 above)
Comment (nickstenning) [inline h/static/scripts/directive/group-list.js]: Ah, understood. Because the "New group" button is different to the rest.

**Comment by nickstenning** [comment]
In the light of @conor's comments, it might make more sense to special-case the "New group" button in the CSS.

**Comment by nickstenning** [comment]
> Should we keep the current style until we're ready to make a uniform shift?

I don't think so. Incremental improvement is good.

File: h/static/scripts/directive/group-list.js [hunk 6]
```diff
@@ -1,17 +1,61 @@
...
+      }
+
+      $scope.$watch('expandedGroupId', function (activeGroupId) {
+        if (activeGroupId) {
+          // wait for the share link field to be revealed and then select
+          // the link's text
+          setTimeout(function() {
+            var activeShareLinkField = elem[0].querySelector('.share-link-field[data-group-id=' + activeGroupId + ']');
+            assert(activeShareLinkField);
```
Comment (nickstenning) [inline h/static/scripts/directive/group-list.js]: Not sure how I feel about assertions in production code. I'd rather we had a `if (!activeShareLinkField) { console.warn(...); }`.

**Comment by conordelahunty** [comment]
> Should we keep the current style until we're ready to make a uniform shift?
> I don't think so. Incremental improvement is good.

Incremental improvement is definitely preferred, but I understand the concerns. The changes I am trying to make will hopefully feel harmonious, but are very much intentional baby steps forward. I can simplify some of the things I am proposing. @dwhly which elements worry you the most?

File: h/static/scripts/directive/group-list.js (hunk 6 above)
Comment (robertknight) [inline h/static/scripts/directive/group-list.js]: The problem with that is that you end up introducing untested code paths. The assert() is redundant here really though since the call to `activeShareLinkField.focus()` would produce a more useful error in this context. I'll remove it.

**Comment by dwhly** [comment]
> Incremental improvement is good.

There are so many places and situations in which that's the case, and I strongly endorse the philosophy generally-- particularly with respect to underlying functionality.  But I *do* think with the UI the application should look and feel consistent.

>  I can simplify some of the things I am proposing. @dwhly which elements worry you the most?

I'm not sure it's simplification per-se that I'm getting at.  It's that these two cards are serving the same type of function with completely different design styles.

Sharing the page (w/ via) and providing social links.

screen shot 2015-09-21 at 9 00 13 am

Sharing a group and providing social links.

screen shot 2015-09-21 at 9 46 44 am

I happen to *like* the look that you're headed towards.  I guess my ask is that we convert things over we do it more or less simultaneously, and put together a dictionary for our overall design vocabulary.

(I'm really not trying to be an obstructionist ass, I swear.)

**Comment by robertknight** [comment]
t87-scope-dropdown

**Comment by conordelahunty** [comment]
@robertknight Thanks Robert, looks tight

File: h/static/styles/app.scss (hunk 3 above)
Comment (tilgovi) [inline h/static/styles/app.scss]: You're going to want to use the Compass mixins for this, or I think you'll miss out on IE10 support.

Comment (tilgovi) [inline h/static/styles/app.scss]: See, for example, https://github.com/hypothesis/h/blob/master/h/static/styles/topbar.scss#L20

**Comment by conordelahunty** [comment]
> (I'm really not trying to be an obstructionist ass, I swear.)

Not at all. It's a fair point. Let me take another quick look and see if we can do something we're both happy with.

File: h/static/styles/app.scss (hunk 3 above)
Comment (robertknight) [inline h/static/styles/app.scss]: > You're going to want to use the Compass mixins for this, or I think you'll miss out on IE10 support.

Ah, indeed. Actually Safari will be the bigger issue, they only dropped the prefix in Safari 9.

Has post-processing with [autoprefixer](https://github.com/postcss/autoprefixer) been considered for handling vendor prefixes automatically?

Comment (tilgovi) [inline h/static/styles/app.scss]: I considered it but wasn't sure it accounted for all the usage.

Comment (tilgovi) [inline h/static/styles/app.scss]: In other words, until we're not using compass anyway, just use compass.

File: h/static/scripts/directive/group-list.js [hunk 7]
```diff
@@ -1,20 +1,66 @@
...
     },
     restrict: 'AE',
     scope: {},
     templateUrl: 'group_list.html'
   };
 };
+
+module.exports = {
+  directive: GroupsListDirective,
+  _Controller: GroupsListController
```
Comment (nickstenning) [inline h/static/scripts/directive/group-list.js]: Not sure about the leading underscore here. If we're exporting we're exporting it, so it should just be `Controller`.

File: h/static/scripts/directive/group-list.js [hunk 8]
```diff
@@ -1,20 +1,66 @@
...
 /**
  * @ngdoc directive
  * @name groupList
  * @restrict AE
  * @description Displays a list of groups of which the user is a member.
  */
-
 // @ngInject
-module.exports = function (groups) {
+function GroupsListDirective(groups, $window) {
```
Comment (nickstenning) [inline h/static/scripts/directive/group-list.js]: The directive is a factory function rather than a strict constructor, so for the sake of consistency I'd say htis should be `groupsList` or similar, rather than `GroupsListDirective`.

Comment (robertknight) [inline h/static/scripts/directive/group-list.js]: Agreed. There are a couple of other mistakes in this commit, I'll ping on Slack when its ready for review.

File: h/static/scripts/directive/group-list.js (hunk 7 above)
Comment (robertknight) [inline h/static/scripts/directive/group-list.js]: I'm using the underscore to denote a 'private' export just for use in the tests, which I think is a fairly common convention.

Comment (nickstenning) [inline h/static/scripts/directive/group-list.js]: I've had this back and forth with Sean a few times in Python land. My view is that if it's private, it's an implementation detail and shouldn't be tested. If it's tested, it's public interface, and shouldn't have its name mangled.

**Comment by conordelahunty** [comment]
@dwhly Hey Dan, took a look at some of the consistency issues we have and put together a little proposal.

I tried to get the group creation, invite, and success/invite others pages look the same consistent. I then tried to get the sharing a document function in the sidebar consistent with the rest of the options. Take a look and let me know what you think.

https://docs.google.com/presentation/d/1Fr681lSNkQCxRUnLsi8WGFocpRQ-n5kaR05dDAwc3lc/edit?usp=sharing

File: h/static/scripts/directive/group-list.js (hunk 7 above)
Comment (robertknight) [inline h/static/scripts/directive/group-list.js]: I think would consider the controller an implementation detail but the advantage of some white-box testing of units is that it makes debugging easier if something fails and the tests run faster (in this case, just some trivial function calls instead of rendering HTML in a browser), so you can have more of them.

I've added tests for the directive now which test the toggle at the component level as well.

File: h/static/scripts/directive/group-list.js [hunk 9]
```diff
@@ -1,20 +1,75 @@
...
 /**
  * @ngdoc directive
  * @name groupList
  * @restrict AE
  * @description Displays a list of groups of which the user is a member.
  */
-
 // @ngInject
-module.exports = function (groups) {
+function groupsList(groups, $window) {
```
Comment (nickstenning) [inline h/static/scripts/directive/group-list.js]: Everywhere else this directive is referred to as `groupList`, so both this and the controller should probably be `group*` (singular) rather than `groups*` (plural).

File: h/static/scripts/directive/group-list.js [hunk 10]
```diff
@@ -1,20 +1,75 @@
...
     },
     restrict: 'AE',
     scope: {},
     templateUrl: 'group_list.html'
   };
 };
+
+module.exports = {
+  directive: groupsList,
+  _Controller: GroupsListController
```
Comment (nickstenning) [inline h/static/scripts/directive/group-list.js]: I hear your explanation, but I'm not convinced. Let's just call this `Controller` and be done with it. There's no harm done by exporting it properly.

File: h/static/scripts/directive/test/privacy-test.js [hunk 11]
```diff
@@ -1,6 +1,6 @@
 'use strict';
 
-var PrivacyController = require('../privacy').PrivacyController;
+var PrivacyController = require('../privacy')._Controller;
```
Comment (nickstenning) [inline h/static/scripts/directive/test/privacy-test.js]: I guess you're changing this in order to be consistent with the changes you've made to the `groupList` directive. That's fine, but ideally we'd keep changes like this in separate commits, if not in an entirely separate PR.

But I'm not a great fan of the change to start with, so perhaps just back this out?

File: h/static/styles/app.scss (hunk 3 above)
Comment (nickstenning) [inline h/static/styles/app.scss]: You still need to address this. For an example, see [this commit](https://github.com/hypothesis/h/commit/3039032).

File: h/static/scripts/directive/group-list.js [hunk 12]
```diff
@@ -1,20 +1,75 @@
...
+    } else {
+      $scope.expandedGroupId = groupId;
+    }
+  };
+
+  $scope.shouldShowShareLink = function (groupId) {
+    return $scope.expandedGroupId === groupId;
+  }
+
+  $scope.sortedGroups = function () {
```
Comment (nickstenning) [inline h/static/scripts/directive/group-list.js]: I'd rather we just sorted them on the server side and had the client render them in the order given.

File: h/static/scripts/directive/test/group-list-test.js [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+// returns true if a jQuery-like element has
+// been hidden directly via an ng-show directive.
+//
+// This does not check whether the element is a descendant
+// of a hidden element
+function isElementHidden(element) {
+  return element.hasClass('ng-hide');
+}
+
+describe('GroupsListDirective', function () {
```
Comment (nickstenning) [inline h/static/scripts/directive/test/group-list-test.js]: `s/GroupsListDirective/groupList/`

File: h/templates/client/group_list.html [hunk 14]
```diff
@@ -2,29 +2,60 @@
...
+        <!-- the group name and share link !-->
+        <div class="group-details">
+          <div class="group-name-container">
+            <a class="group-name-link"
+               href=""
+               title="{{ group.public ? 'Show public annotations' : 'Show and create annotations in ' + group.name }}">
+               {{group.name}}
+            </a>
+          </div>
+          <div class="share-link-container" ng-click="stopClickPropagation($event)" ng-if="!group.public">
```
Comment (nickstenning) [inline h/templates/client/group_list.html]: Not sure the indirection of `stopClickPropagation` adds anything here. `ng-click="$event.stopPropagation()"` would achieve the same thing and is equally clear, I think.

File: h/static/scripts/directive/group-list.js (hunk 12 above)
Comment (robertknight) [inline h/static/scripts/directive/group-list.js]: > If you need a hand making that change, just shout.

Yes - I haven't looked into the backend that serves the API yet at all having a few pointers would be useful.

File: h/templates/client/group_list.html (hunk 14 above)
Comment (robertknight) [inline h/templates/client/group_list.html]: Agreed - done.

File: h/static/scripts/directive/test/group-list-test.js (hunk 13 above)
Comment (robertknight) [inline h/static/scripts/directive/test/group-list-test.js]: Done

File: h/static/scripts/directive/group-list.js (hunk 10 above)
Comment (robertknight) [inline h/static/scripts/directive/group-list.js]: Done

File: h/static/styles/app.scss (hunk 3 above)
Comment (robertknight) [inline h/static/styles/app.scss]: @tilgovi - Is there a previous discussion about removing compass somewhere or does anyone else know what needs to be done? Ideally I'd like all the client deps to be handled via npm

Comment (nickstenning) [inline h/static/styles/app.scss]: https://github.com/hypothesis/h/issues/1292

**Comment by nickstenning** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;08dCx7gd&#x2F;87-group-scope-dropdown-ui-updates"> Group scope dropdown UI updates</a>

**Comment by nickstenning** [comment]
The only thing blocking merge here is updates to the CSS. Let me know if I'm out of date on that.

**Comment by robertknight** [comment]
@nickstenning - That's correct and the CSS issue has been addressed.

**Comment by nickstenning** [comment]
LGTM. Nice work. :cake:

**Comment by lenazun** [comment]
🎉
